### ISSUE TITLE
Support test of private classes and defines

### ISSUE THREAD

**Issue by ITler** [rodjek/rspec-puppet#325]
The puppetlabs stdlib library supports [marking classes and defines as private](https://github.com/puppetlabs/puppetlabs-stdlib#assert_private), so that they can only be accessed within their module. Yet, It seems, I can't test these classes and defines with _rspec-puppet_ as they get accessed from the outside.

Are there plans to support this?

**Comment by DavidS** [comment]
Did you try stubbing out the private function? It's probably something to have an expectation for anyways?

**Comment by ITler** [comment]
Again, I've learned something new. I've taken [this](https://github.com/rodjek/rspec-puppet/issues/102#issuecomment-35568335) as a template for my mock.
I agree having an expectation would make sense.

(By the way, I wonder, why I am the first one asking for that. Maybe there is a more simple and convenient way to get this issue solved?)

What would be the best way (which means easy and centralized implementation) to provide this mock to other spec files within and outside my current module?

**Comment by DavidS** [comment]
There is an example of function stubbing in https://github.com/rodjek/rspec-puppet#accessing-the-parser-scope-where-the-function-is-running ; you can use any stubbing/mocking framework you prefer on the `scope` object.

To provide the stub to multiple tests, extracting it into a shared context of RSpec: https://www.relishapp.com/rspec/rspec-core/docs/example-groups/shared-context

(I guess `private` is not a very popular feature.)

**Comment by ITler** [comment]
@DavidS Thanks, this helps... Maybe we can leave this issue open for implementing the expectation of a function to be called?

**Comment by DavidS** [comment]
If you use "expects", the test should fail if the function is not called. If you just want to "override" the function, use "stubs". That will not fail when the method is not called. This is part of rspec itself (See the allowing and expecting messages sections in https://relishapp.com/rspec/rspec-mocks/v/3-1/docs/basics). I think the syntax in the respec-puppet README is mocha (https://rubygems.org/gems/mocha), that seems to be being pulled in by something else - I've only very recently noticed that discrepancy.

**Comment by ITler** [comment]
Yeah, I think I understand. I already use _stubbing_ to override the [`assert_private`](https://github.com/puppetlabs/puppetlabs-stdlib#assert_private) function using shared context as you proposed. This works fine for me. 

Would it make sense to implement an expectation for a function call? Maybe it does for function with `:type => :statement`, only?
Example:

    describe 'foo', type: :class do
      context 'with defaults for all parameters' do
        it { should call_function(:assert_private) } # green if: class foo { assert_private() }
        # or if function provides parameter
        it { should call_function(:any_statement_function_with_parameter).with_params(['a', 'b']) }
      end
    end

**Comment by DavidS** [comment]
Ah, now I understand what you're saying. It is hard to fit the usual mock expectations into the call flow of an rspec-puppet run. Mainly because the catalog compilation happens semi-non-deterministically while evaluating the specs and has its results cached. :-/ I don't see a quick fix for that.

**Comment by ITler** [comment]
If I correctly understand your last comment, this means, determining presence of a class definition within a puppet manifest (by using `contain_class`) is _easier_ (thus already implemented) than determining if a function gets called.
Well, I assume, this issue never gets realized in Puppet as long as catalogue compilation method stays same? Will it make sense to let this issue open?

**Comment by DavidS** [comment]
Yes, exactly. That's why I removed the "awaiting feedback" label and kept it open: being able to **easily** add expectations around function calls during catalog compilation is something that would be nice to have.

**Comment by ITler** [comment]
Great

**Comment by cdenneen** [comment]
@ITler @DavidS Any further traction on this? I'd like to test some private classes to but am getting my "Class is private" error.

**Comment by baurmatt** [comment]
For documentation... This is needed to mock the assert_private() function:

```
  before(:each) do
    # Fake assert_private function from stdlib to not fail within this test
    Puppet::Parser::Functions.newfunction(:assert_private, :type => :rvalue) { |args|
    }
  end
```

**Comment by tthayer** [comment]
@baurmatt That's some black magic right there. Thank you. We've been trying to mock `Puppet::Functions.create_function()` stuff for months without making any headway.

**Comment by philhod** [comment]
hi,
is there anyway to allow the mocked function to take a block? It fails with an Argument Error 'does not expect a block' and I've tried various spellings which don't compile, e.g.
let(:pre_condition) do
  'function i_take_a_block () { yield 1 }' => Arg error 'does not expect a block'
end
let(:pre_condition) do
  'function i_take_a_block (&my_block) { my_block.call(1) }' => compile fail
end

**Comment by baurmatt** [comment]
`pre_condition` uses Puppet code, not Ruby.

**Comment by traylenator** [comment]
The following does seem to work and allows you keep your mocks in ruby.

```ruby
before(:each) do
    # Fake assert_private function from stdlib to not fail within this test
    Puppet::Parser::Functions.newfunction(:assert_private, :type => :rvalue) { |args|
    }
end
```

becomes

```ruby
before(:each) do
    # Fake assert_private function from stdlib to not fail within this test
    Puppet::Parser::Functions.newfunction(:my_assert_private, :type => :rvalue) { |args|
    }
  end
```
```ruby
let(:pre_condition) do
   'function assert_private($foo) { return my_assert_private($foo) }'
end
```

**Comment by ITler** [comment]
I think this can be closed then
